### PR TITLE
Fix linking of generated files in cmake

### PR DESCRIPTION
**Pull request by paul-elliott-arm**
## Description

If generation of files is turned off, and a file is missing, when building in tree with cmake, you can end up with the generated file being turned into a symlink to itself. This will also break any future attempt at building with make. Fix this by testing if the file exists prior to attempting to link it.

## Status
**READY**

## Requires Backporting
NO  

## Migrations
NO

## Additional comments
This was found by (and contributed to) the breakage of the recent release.

## Todos
- [x] Tests
- [x] Changelog updated

## Steps to test or reproduce
CMake both in and out of tree should work with GEN_FILES off and although it should fail with missing files, should not create a self referential symlink.

### CODE REVIEW

**Review by d3zd3z — COMMENTED** [review]
Approved

**Comment by daverodgman** [comment]
Fails for me with:

`make clean

(pr6086) 16:28:18 ~/code/mbedtls/mbedtls$ cmake  -DGEN_FILES=0 .
CMake Error at CMakeLists.txt:150 (message):
  Could not create symbolic link for:
  /Users/davrod01/code/mbedtls/mbedtls/programs/test/query_config.c --> ln:
  /Users/davrod01/code/mbedtls/mbedtls/programs/test/query_config.c: File
  exists

Call Stack (most recent call first):
  programs/test/CMakeLists.txt:59 (link_to_source)


-- Configuring incomplete, errors occurred!
See also "/Users/davrod01/code/mbedtls/mbedtls/CMakeFiles/CMakeOutput.log".
See also "/Users/davrod01/code/mbedtls/mbedtls/CMakeFiles/CMakeError.log".
(pr6086) 16:28:24 ~/code/mbedtls/mbedtls$ ls -ls /Users/davrod01/code/mbedtls/mbedtls/programs/psa/psa_constant_names_generated.c
0 lrwxr-xr-x  1 davrod01  staff  80 19 Jul 16:28 /Users/davrod01/code/mbedtls/mbedtls/programs/psa/psa_constant_names_generated.c -> /Users/davrod01/code/mbedtls/mbedtls/programs/psa/psa_constant_names_generated.c`

File: library/CMakeLists.txt [hunk 1]
```diff
@@ -173,10 +173,18 @@ if(GEN_FILES)
 
 
 else()
-    link_to_source(error.c)
+    if(EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/error.c)
```
[the remark this answers is not in the document]
Comment (paul-elliott-arm) [reply library/CMakeLists.txt:176]: I was concerned that EXISTS() can be rather sensitive when it comes to not having exact paths specified, however after fairly extensive testing, this seems to not be an issue.

**Comment by paul-elliott-arm** [comment]
> Fails for me with:
> 
> `make clean
> 
> (pr6086) 16:28:18 ~/code/mbedtls/mbedtls$ cmake -DGEN_FILES=0 . CMake Error at CMakeLists.txt:150 (message): Could not create symbolic link for: /Users/davrod01/code/mbedtls/mbedtls/programs/test/query_config.c --> ln: /Users/davrod01/code/mbedtls/mbedtls/programs/test/query_config.c: File exists
> 
> Call Stack (most recent call first): programs/test/CMakeLists.txt:59 (link_to_source)
> 
> -- Configuring incomplete, errors occurred! See also "/Users/davrod01/code/mbedtls/mbedtls/CMakeFiles/CMakeOutput.log". See also "/Users/davrod01/code/mbedtls/mbedtls/CMakeFiles/CMakeError.log". (pr6086) 16:28:24 ~/code/mbedtls/mbedtls$ ls -ls /Users/davrod01/code/mbedtls/mbedtls/programs/psa/psa_constant_names_generated.c 0 lrwxr-xr-x 1 davrod01 staff 80 19 Jul 16:28 /Users/davrod01/code/mbedtls/mbedtls/programs/psa/psa_constant_names_generated.c -> /Users/davrod01/code/mbedtls/mbedtls/programs/psa/psa_constant_names_generated.c`

Fixed via the moving of the check into` link_to_source()`

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

File: CMakeLists.txt [hunk 2]
```diff
@@ -131,7 +131,10 @@ function(link_to_source base_name)
         set(target "${CMAKE_CURRENT_SOURCE_DIR}/${base_name}")
     endif()
 
-    if (NOT EXISTS ${link})
+    # Linking to non-existant file is not desirable. At best you will have a
```
Comment (daverodgman) [inline CMakeLists.txt]: Spelling: s/existant/existent

**Review by daverodgman — APPROVED** [review]
LGTM

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Pull request merged**